### PR TITLE
Add initial Elasticsearch mappings

### PR DESCRIPTION
**Pull request by kenoir**
## What is this PR trying to achieve?

In order to only ingest valid data, and to optimise our indexes for search, we should have Elasticsearch mappings defined.

In this first pass we are looking to accommodate Miro data.

## Who is this change for?

Anyone who likes structured data.

## Have the following been considered/are they needed?

- [x] Tests?
- [x] Docs?
- [x] Spoken to the right people?

### CODE REVIEW

**Comment by kenoir** [comment]
`elastic4s` seems under rapid development so finding the right way to define mapping involves ensuring you've pinned the repo at the correct version and sometimes trawling the code.

File: common/src/main/scala/uk/ac/wellcome/lib/elasticsearch/mappings.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package uk.ac.wellcome.lib.elasticsearch
+
+import com.sksamuel.elastic4s._
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s.mappings.FieldType._
+import com.sksamuel.elastic4s.analyzers._
+
+// TODO: This should have a test before any more code is written.
```
Comment (kenoir) [inline common/src/main/scala/uk/ac/wellcome/lib/elasticsearch/mappings.scala]: Sorry ... should have written the tests first. I wasn't sure if this was even possible for our purposes with the elastic4s library, but it seems it is.

**Comment by kenoir** [comment]
We'll want to add created and updated datetime fields for the doc is ES too.

File: common/src/main/scala/uk/ac/wellcome/elasticsearch/mappings/RecordsIndex.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package uk.ac.wellcome.elasticsearch.mappings
+
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s._
+import com.sksamuel.elastic4s.analyzers._
+import com.sksamuel.elastic4s.mappings.DynamicMapping
+
+class RecordsIndex(client: ElasticClient, indexName: String) {
```
Comment (kenoir) [inline common/src/main/scala/uk/ac/wellcome/elasticsearch/mappings/RecordsIndex.scala]: We should test at the very least that this creates an index appropriately named.

**Comment by alicefuzier** [comment]
@kenoir Yeah, about tests for the RecordsIndex I wasn't quite sure what to test cause testing that after the creation it has an index with those properties seemed kind of pointless. 
All tests against elasticsearch use the ElasticSearchLocal class so I added a call to our mapping scrpit in there, so that each test in the ingestor or the api runs with an elasticsearch provisioned that way. 
This tests both that an index exist with that name and that the APi and the ingestor can use that index, so it's actually a greater coverage. 
I did add a RecordsIndexTest class in the end cause there was no other way to test if elasticsearch would reject an item in the wrong format

File: common/src/main/scala/uk/ac/wellcome/elasticsearch/mappings/RecordsIndex.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package uk.ac.wellcome.elasticsearch.mappings
+
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s._
+import com.sksamuel.elastic4s.analyzers._
+import com.sksamuel.elastic4s.mappings.DynamicMapping
+
+class RecordsIndex(client: ElasticClient, indexName: String, itemType: String) {
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/elasticsearch/mappings/RecordsIndex.scala]: Maybe call this "WorksIndex"?

File: common/src/test/scala/uk/ac/wellcome/elasticsearch/mappings/WorksIndexTest.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.elasticsearch.transport.RemoteTransportException
+import org.scalatest.FunSpec
+import org.scalatest.concurrent.ScalaFutures
+import uk.ac.wellcome.models.{IdentifiedWork, SourceIdentifier, Work}
+import uk.ac.wellcome.test.utils.ElasticSearchLocal
+import uk.ac.wellcome.utils.GlobalExecutionContext.context
+import uk.ac.wellcome.utils.JsonUtil
+
+class WorksIndexTest extends FunSpec with ElasticSearchLocal with ScalaFutures{
+  // RecordsIndex is actually used in ElasticSearchLocal to create the index.
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/elasticsearch/mappings/WorksIndexTest.scala]: comment refers to the old name

**Comment by kenoir** [comment]
Looks good.

**Pull request merged**